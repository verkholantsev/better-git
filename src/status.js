// @flow

import mapOpts from './map-opts';

import type { Git } from './git-factory';
import type { Opts } from './map-opts';

type Status = 'renamed' | 'deleted' | 'modified' | 'new' | 'added';

export type FileStatus = {
    destination?: string,
    isStaged: boolean,
    name?: string,
    source?: string,
    status: Status,
};

export default async function status(git: Git, opts: Opts = {}): Promise<Array<FileStatus>> {
    const args = ['status', '--porcelain', ...mapOpts(opts)];
    const out = await git(args);
    return parseStatus(out);
}

const RENAMED_REGEX = /^R\s(.+)\s->\s(.+)$/;
const STATUS_REGEX = /^\s*[A-Z?]+\s+/;

function parseStatus(inputText: string): Array<FileStatus> {
    const lines = inputText.split('\n').filter(Boolean);

    return lines.map(line => {
        if (line.startsWith('R ')) {
            const groups = RENAMED_REGEX.exec(line);

            if (!groups) {
                throw new Error(`Unexpected line in 'git status --porcelain' output '${line}`);
            }

            const [, source, destination] = groups;
            return { source, destination, status: 'renamed', isStaged: true };
        } else if (line.startsWith(' D') || line.startsWith('D ')) {
            const isStaged = line.startsWith('D ');
            const name = line.replace(STATUS_REGEX, '');
            return { name, status: 'deleted', isStaged };
        } else if (line.startsWith(' M') || line.startsWith('M ')) {
            const isStaged = line.startsWith('M ');
            const name = line.replace(STATUS_REGEX, '');
            return { name, status: 'modified', isStaged };
        } else if (line.startsWith('A ')) {
            const name = line.replace(STATUS_REGEX, '');
            return { name, status: 'added', isStaged: true };
        } else if (line.startsWith('??')) {
            const name = line.replace(STATUS_REGEX, '');
            return { name, status: 'new', isStaged: false };
        }

        throw new Error(`Unexpected line in 'git status --porcelain' output '${line}'`);
    });
}
