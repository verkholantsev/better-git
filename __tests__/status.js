// @flow

import fs from 'fs';
import path from 'path';
import promisify from 'util.promisify';

import status from '../src/status';

const readFile = promisify(fs.readFile);

describe('git.status()', () => {
    it('should work correctly', async () => {
        const fixturePath = path.resolve(__dirname, '__fixtures__', 'status-porcelain');
        const git = async () => await readFile(fixturePath, { encoding: 'utf8' });

        expect(await status(git)).toMatchSnapshot();
    });
});
