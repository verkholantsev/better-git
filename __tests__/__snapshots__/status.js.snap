// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git.status() should work correctly 1`] = `
Array [
  Object {
    "destination": "__tests__/integration.1.js",
    "isStaged": true,
    "source": " __tests__/integration.js",
    "status": "renamed",
  },
  Object {
    "isStaged": false,
    "name": "flow-typed/npm/flow-copy-source_vx.x.x.js",
    "status": "deleted",
  },
  Object {
    "isStaged": false,
    "name": "src/parse-commits.js",
    "status": "modified",
  },
  Object {
    "isStaged": true,
    "name": "src/status.js",
    "status": "added",
  },
  Object {
    "isStaged": false,
    "name": "__tests__/__fixtures__/status-porcelain",
    "status": "new",
  },
  Object {
    "isStaged": false,
    "name": "__tests__/status.js",
    "status": "new",
  },
]
`;
